subsys/profiler/profiler_nordic.c

```diff
@@ -241,12 +241,29 @@ uint16_t profiler_register_event_type(const char *name, const char * const *args
 	return ne;
 }
 
+void profiler_log_encode_VLQ(struct log_event_buf *buf, uint32_t data)
+{
+	/* Encoding using Variable-length quantity (VLQ) */
+	while (data > 127) {
+		profiler_log_encode_uint8(buf, (uint8_t) data | 0x80);
+		data = data >> 7;
+	}
+	profiler_log_encode_uint8(buf, (uint8_t) data);
+}
+
 void profiler_log_start(struct log_event_buf *buf)
 {
+	static uint32_t old_time;
 	/* Adding one to pointer to make space for event type ID */
 	__ASSERT_NO_MSG(sizeof(uint8_t) <= CONFIG_PROFILER_CUSTOM_EVENT_BUF_LEN);
 	buf->payload = buf->payload_start + sizeof(uint8_t);
-	profiler_log_encode_uint32(buf, k_cycle_get_32());
+	int key = irq_lock();
+
+	uint32_t new_time = k_cycle_get_32();
+
+	profiler_log_encode_VLQ(buf, (uint32_t) new_time - old_time);
+	irq_unlock(key);
+	old_time = new_time;
```

Comment:
But there is more to it then getting the timestamps in order. Note that messages must also arrive in order and if you suspect that this function could race, you will have the same race with sending the packet.
Can it be fixed... possibly but there will be multiple problems with it. First you must allocate the message buffer at the same time you do the timestamp. Second you must be sure that no message is dropped and if it is there is a way to recover timestamps. Once I suggested to send a special message that will be a time reference and all messages will reference to it. But it will have issues too.

So we can compress the timestamp the way you proposed if you still think it has sense.